server/src-lib/Hasura/Db.hs

```diff
@@ -56,28 +58,28 @@ instance (MonadTx m) => MonadTx (ValidateT e m) where
 data LazyTx e a
   = LTErr !e
   | LTNoTx !a
-  | LTTx !(Q.TxE e a)
+  | LTTx !(Q.TxE e a) !(Maybe Q.TxAccess)
 
 lazyTxToQTx :: LazyTx e a -> Q.TxE e a
 lazyTxToQTx = \case
   LTErr e  -> throwError e
   LTNoTx r -> return r
-  LTTx tx  -> tx
+  LTTx tx _txaM -> tx
 
 runLazyTx
```

Comment:
The other idea (and simpler IMO) was to provide the access mode as a parameter to `runLazyTx` but metadata queries are all peeled outside the main handlers and hence don't use `runLazyTx` at all, if I understand correctly. See: https://github.com/tirumaraiselvan/graphql-engine/blob/e6c83ac646ed16a31667a78a3ba98bccf4e7a70e/server/src-lib/Hasura/Server/Query.hs#L201